### PR TITLE
Apply packmol docs suggestions for box size with PBC

### PR DESCRIPTION
**Pull request by andrrizzi**
Packmol does not support periodic boundary conditions. Their [docs suggests](http://www.ime.unicamp.br/~martinez/packmol/userguide.shtml#pbc) to pack the molecules in a box 2 angstroms smaller.

I've also made a modification to `openmoltools.pack_box` so that all temporary files created by the function are deleted at the end of the execution (`tempfile.mktemp` does not delete them automatically).

### CODE REVIEW

File: openmoltools/packmol.py [hunk 1]
```diff
@@ -121,52 +122,60 @@ def pack_box(pdb_filenames_or_trajectories, n_molecules_list, tolerance=2.0, box
...
+
+        # Approximating volume to initialize box.
+        if box_size is None:
+            box_size = approximate_volume(pdb_filenames, n_molecules_list)
+
+        # Adjust box_size for periodic box. Packmol does not explicitly
+        # support periodic boundary conditions and the suggestion on
+        # theri docs is to pack in a box 2 angstroms smaller. See
+        # http://www.ime.unicamp.br/~martinez/packmol/userguide.shtml#pbc
+        packmol_box_size = box_size - 2  # angstroms
```
Comment (davidlmobley) [inline openmoltools/packmol.py]: My one concern with this specific implementation is that for particularly small and dense boxes, this might actually make it so packmol can't find a solution. Have you checked, @andrrizzi ? If that's the case then you would need to also modify utilities which guess the box size (`approximate_volume`, `approximate_volume_bydensity`) so that they make a corresponding increase in the target box size (so that you're ending up with the same size solvent box we currently have, plus 2 angstroms of additional buffer, rather than with a smaller solvent box than we currently have plus 2 angstroms of buffer so the box size is the same as we currently have.)

You should be able to check this, I think, by solvating, say, methane in cyclohexane based on density (with a guessed density matching that of water) since this should result in a rather small box that is overly packed.

**Comment by davidlmobley** [comment]
Thanks, @andrrizzi . I added one concern/aspect you might want to test but otherwise it looks good to me.

File: openmoltools/packmol.py (hunk 1 above)
Comment (jchodera) [inline openmoltools/packmol.py]: Instead of decreasing the box size by 2 A before the packmol step, I would instead *increase* the box size by 2 angstroms following the packmol step.

**Comment by andrrizzi** [comment]
Thanks for suggestion @davidlmobley! I'll check it out.

File: openmoltools/packmol.py (hunk 1 above)
Comment (andrrizzi) [inline openmoltools/packmol.py]: Hmm. Sorry, I'm not sure I follow. If I increased the box size used by packmol by 2A, I'd have to increase the `Trajectory` box size by 4A to make simulations work, right?

File: openmoltools/packmol.py [hunk 2]
```diff
@@ -121,52 +122,60 @@ def pack_box(pdb_filenames_or_trajectories, n_molecules_list, tolerance=2.0, box
...
+            filename = pdb_filenames[k]
+            n_molecules = n_molecules_list[k]
+            header += BOX_TEMPLATE % (filename, n_molecules, packmol_box_size,
+                                      packmol_box_size, packmol_box_size)
+
+        print(header)
+
+        packmol_filename = tempfile.mktemp(suffix='.txt', dir=tmp_dir)
+        with open(packmol_filename, 'w') as file_handle:
+            file_handle.write(header)
```
Comment (davidlmobley) [inline openmoltools/packmol.py]: Wait, don't you have to also add back on the 2 angstroms you subtracted off after you finish passing through packmol?

**Comment by jchodera** [comment]
I just spoke to @andrrizzi, and I think the only change we should be making in this PR is to add 2A to each box dimension after packmol has created the packed geometry. @andrrizzi was somewhat concerned about @davidlmobley doing NVT simulations after this, but since they always do NPT equilibration first, it should still have ample time to recover the appropriate density.

File: openmoltools/packmol.py (hunk 1 above)
Comment (davidlmobley) [inline openmoltools/packmol.py]: I think we're both saying variants of the same thing: Your current approach results in using a box size that is 2 angstroms smaller than intended throughout, which is not quite what we want. 

I think you either want:
a) John's approach: Pack into the target box size, then increase the box size by 2 angstroms after packing into it
b) My approach: Pack into a box size 2 angstroms smaller than the target, then increase the box size by 2 angstroms after packing into it 

Either can work. But approach (a) will require you changing the behavior of `pack_box` when `box_size` is provided (and updating doc strings) so that if the user asks for a 4 nm box, you don't give them back a 4.2 nm box (which is what approach (a) would do). Approach (b) would leave the behavior of `pack_box` the same when the user provides `box_size` (by packing into a box 2 angstroms smaller than expanding it by 2 angstroms) and would fix the "overly dense box" issue by modifying the functions `approximate_volume...` to give box edge estimates which are 2 angstroms larger than they need to be, so you have room to (i) shrink by 2 angstroms, (ii) pack, and (iii) increase by 2 angstroms.

So basically I think you need to do (a) or (b) and I'm OK with either. But your current approach will probably fail for boxes which are too dense, AND results in final box sizes which still don't include the 2 angstrom buffer unless I'm missing where that's done.

**Comment by davidlmobley** [comment]
@jchodera - see my point above about options (a) and (b). Your approach is fine with me, but it will require some attention paid to the case where the user has requested a specific box size, since adding two angstroms would then result in the final box being larger than the user requested.

File: openmoltools/packmol.py (hunk 1 above)
Comment (jchodera) [inline openmoltools/packmol.py]: Isn't the problem with (a) exactly what you highlighted above?
> My one concern with this specific implementation is that for particularly small and dense boxes, this might actually make it so packmol can't find a solution.

Comment (davidlmobley) [inline openmoltools/packmol.py]: No. (a) and (b) are both solutions that will work, I think. I'm just saying that doing what Andrea INITIALLY did (just shrink the box size from whatever it was) may not work in some cases.

Comment (jchodera) [inline openmoltools/packmol.py]: Ah, OK. I'm fine with solution (a) then if it won't cause packmol to sputter and die. I was worried that it would make it nearly impossible for packmol to pack very dense boxes because we're not also scaling the rejection overlap cutoff.

File: openmoltools/packmol.py (hunk 2 above)
Comment (andrrizzi) [inline openmoltools/packmol.py]: That is a temporary file. The 2 angstroms are added back to the trajectory that is returned by the function.

**Comment by andrrizzi** [comment]
Currently option (b) is implemented. The 2 angstroms are correctly added back into the `Trajectory` object (note that I'm setting `trj.unitcell_vectors` using `box_size`, which is different than the decreased version passed to packmol `packmol_box_size`).

I'd be in favor of making `approximate_volume_by_density` add 2A to the returned box size, so that the box size returned by `pack_box` will be the same as the one passed by the argument. That function doesn't return the theoretical volume anyway, since it already scales it up by a 1.1 factor.

I can add an extra flag to the function `approximate_volume_by_density(..., box_buffer=2.0)` in case we want to control, and to make sure that users are aware of it.

**Comment by davidlmobley** [comment]
That plan sounds great, @andrrizzi .

**Comment by andrrizzi** [comment]
Actually, I run the whole `SolvationToolkit` test suite with the current implementation and the tests pass. I also tried to solvate 1 molecule of methane in 300 of cyclohexane with water density and packmol was able to find a solution (although it took 10s longer).

@davidlmobley do you still think we should add the 2A box buffer to the `box_size` estimation given by `approximate_volume_by_density` for safety?

**Comment by davidlmobley** [comment]
10s longer is pretty significant, right? (I recall this being just a matter of a second or two normally...?) If so that'd be an argument to increasing it. But otherwise (if it's 10s out of a couple of minutes) then I'm OK without increasing, @andrrizzi .

**Comment by andrrizzi** [comment]
It's 22s instead of 12s, so almost twice the time.

**Comment by andrrizzi** [comment]
However, the whole `SolvationToolkit` test suite runs in the same time, so the slow down may only affect very tightly packed systems.

**Comment by andrrizzi** [comment]
Done! @davidlmobley @jchodera , ok if I go ahead with the release?

**Comment by davidlmobley** [comment]
Yes, please, @andrrizzi . Thanks.
